File: src/test/java/com/example/demo/NotesUnitTest.java
package com.example.demo;

import com.example.demo.cotrollers.NotesController;
import com.example.demo.models.Note;
import com.example.demo.repositories.NoteRepository;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.Mockito.*;

public class NotesUnitTest {

    @InjectMocks
    private NotesController noteController;

    @Mock
    private NoteRepository noteRepository;

    @BeforeEach
    public void setUp() {
        MockitoAnnotations.openMocks(this); // Inicializa los mocks
    }

    @Test
    public void testHiEndpoint() {
        // Arrange
        NotesController controller = new NotesController();

        // Act
        String response = controller.hi();

        // Assert
        assertEquals("Hello, World!", response);
    }

    @Test
    public void testAddNote() {
        // Arrange: Crea una nota de prueba
        Note note = new Note();
        note.setText("Test Note");

        // Configure the emuated repository
        when(noteRepository.save(note)).thenReturn(note); // Devuelve la misma nota


        Note createdNote = noteController.addNote(note);

        // Assert: Verifica que se haya guardado correctamente
        assertEquals("Test Note", createdNote.getText());
        verify(noteRepository, times(1)).save(note); // Verifica que se llamó al método save
    }
}

File: src/test/java/com/example/demo/NotesUnitTest.kt
import com.example.demo.models.Note
import com.example.demo.controllers.NotesController
import com.example.demo.cotrollers.TestController
import com.example.demo.sevices.NotesService
import com.example.demo.repositories.NoteRepository

import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertNotNull

import org.mockito.Mock
import org.mockito.InjectMocks
import org.mockito.Mockito.mock

import org.mockito.kotlin.verify
import org.mockito.kotlin.times
import org.mockito.Mockito.`when`
import org.mockito.MockitoAnnotations

class NotesTest {

    @Mock
    lateinit var noteRepository: NoteRepository

    @Mock
    lateinit var notesService: NotesService
    @Mock
    lateinit var testController: TestController

    lateinit var notesController: NotesController

    private lateinit var closeable: AutoCloseable

    @BeforeEach
    fun setUp() {
        closeable = MockitoAnnotations.openMocks(this)
        // Crear la instancia de NotesController pasando los mocks
        notesController = NotesController(noteRepository, notesService)
    }

    @AfterEach
    fun tearDown() {
        closeable.close()
    }

    @Test
    fun testHiEndpoint() {
        // Crear mock del repositorio (esto ya lo hace @Mock y @InjectMocks)
        // No es necesario crear manualmente el controller, gracias a @InjectMocks
        // pero si quieres hacerlo manualmente:

        val response = testController.hi()
        assertEquals("Hello, World!", response)
    }


    @Test
    fun testAddNote() {
        val note = Note(id = null, text = "Test Note")
        val savedNote = Note(id = 1L, text = "Test Note")

        // Mockea noteRepository.save() para devolver savedNote
        `when`(noteRepository.save(note)).thenReturn(savedNote)


        assertNotNull(notesController)
        println(notesController.noteRepository)
        // Llamada al método del controlador
        val result = notesController.addNote(note)

        // Verificación
        assertEquals("Test Note", result.text)

        // Verificamos que se llamó a save en el repositorio
        verify(noteRepository, times(1)).save(note)

        // También puedes verificar que se llamó a saveNote en el servicio
        verify(notesService, times(1)).saveNote(note)
    }
}